### PR TITLE
Build macOS engine as an xcframework

### PR DESCRIPTION
**Pull request by vashworth**
Creates and adds FlutterMacOS.xcframework to out/mac or out/host.

Creates and archives FlutterMacOS.xcframework when building mac_host_engine. Archives the xcframework in a new zipped folder at `darwin-x64/framework.zip`, `darwin-x64-profile/framework.zip`, `darwin-x64-release/framework.zip`.

The FlutterMacOS.framework is also still archived currently - I thought it'd be better to keep it archived so we don't have to worry about the tool breaking until we're ready to remove it.

Part of https://github.com/flutter/flutter/issues/126016.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by vashworth** [comment]
@jmagman I'm not sure how to test this. When you did it for https://github.com/flutter/engine/pull/22506, seems like you tested it by adding it as a framework in the scenarios app. However, macos doesn't have a scenarios app

**Comment by jmagman** [comment]
> When you did it for #22506, seems like you tested it by adding it as a framework in the scenarios app. However, macos doesn't have a scenarios app

It's not ideal, but you could add tests in the framework, ready to merge right after the commit rolls into the framework and you are pretty confident it won't be reverted (probably not worth a manual roll).

https://github.com/flutter/flutter/blob/83aca58fa7cdc4b8ba019cbad863afbef11c6c05/packages/flutter_tools/test/host_cross_arch.shard/ios_content_validation_test.dart#L32-L53

Note that the framework codesigning tests run on beta and stable will fail until you add `FlutterMacOS.xcframework`:

https://github.com/flutter/flutter/blob/83aca58fa7cdc4b8ba019cbad863afbef11c6c05/dev/bots/test.dart#L1729
https://github.com/flutter/flutter/blob/83aca58fa7cdc4b8ba019cbad863afbef11c6c05/dev/bots/test.dart#L1759

File: sky/tools/create_macos_framework.py
Comment (jmagman) [inline sky/tools/create_macos_framework.py:239]: Why is there a zip inside the xcframework?

Comment (vashworth) [reply sky/tools/create_macos_framework.py:239]: Oops, that was a copy mistake. Good catch

**Comment by jmagman** [comment]
> The FlutterMacOS.framework is also still archived currently - I thought it'd be better to keep it archived so we don't have to worry about the tool breaking until we're ready to remove it.

+1, let's make sure we have the tool change as ready as possible so the artifact size isn't doubled on the next stable.

**Comment by vashworth** [comment]
Corresponding framework PR: https://github.com/flutter/flutter/pull/143244

**Comment by vashworth** [comment]
Pausing work on this temporarily, though.

**Comment by vashworth** [comment]
This PR should not affect `verify_binaries_codesigned` test or the tool.

`verify_binaries_codesigned` uses artifacts downloaded by the tool
https://github.com/flutter/flutter/blob/c530276f7806c77da2541c518a0e103c9bb44f10/dev/bots/test.dart#L1660-L1667.

And the tool only downloads artifacts specifically specified to be downloaded. For example: https://github.com/flutter/flutter/blob/c530276f7806c77da2541c518a0e103c9bb44f10/packages/flutter_tools/lib/src/flutter_cache.dart#L860-L869

From what I can tell, uploading this new zip file shouldn't affect anything.

**Comment by vashworth** [comment]
@zanderso Would you be comfortable reviewing? @jmagman is OOO. Also, just realized there's a merge conflict, I'll fix it shortly 😅

File: shell/platform/darwin/macos/BUILD.gn
Comment (zanderso) [inline shell/platform/darwin/macos/BUILD.gn:312]: nit:
```suggestion
    rebase_path(_flutter_framework_dir),
```

File: shell/platform/darwin/macos/BUILD.gn
Comment (zanderso) [inline shell/platform/darwin/macos/BUILD.gn:316]: nit:
```suggestion
    rebase_path(root_out_dir),
```

**Review by zanderso — APPROVED** [review]
lgtm

**Pull request merged**
